proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1030 @@
+    data Str
+    null   :: Str
+    append :: Str -> Str -> Str
+
+Other modules in the library import this signature and can use
+the types and functions declared in the signature as if it were
+an ordinary module (i.e., *just normal Haskell*, with *no copy-pasting*)::
+
+    -- in Concat.hs
+    module Concat where
+    import Str
+    concat xs = foldr append null xs
+
+Signatures are declared in the Cabal file as a ``required-signature``
+and parametrize *all* of the modules in the library (providing
+*package-level modularity*)::
+
+    library concat-indef
+        required-signatures: Str
+        exposed-modules: Concat
+
+A library with required signatures is called an **indefinite library**.
+As it is missing implementations for its signatures, it cannot be
+compiled; however, it can still be type-checked (*separate
+type-checking*) and registered with the compiler, so that it can be used
+by other indefinite libraries which depend on it.
+
+An indefinite library can be used in two ways:
+
+1. It can be depended upon by other libraries, which inherit
+   the requirement of this library::
+
+        -- in the Cabal file
+        library stringutils-indef
+            build-depends: concat-indef
+            exposed-modules: StringUtils
+
+        -- in StringUtils.hs
+        module StringUtils where
+        import Concat
+        import Str -- the signature is importable too
+
+   An inherited requirement can be extended simply by adding
+   a local signature.  Similarly, if two dependencies have
+   the same requirement, the two requirements are merged
+   together (*interface composibility.*)
+
+   Using these facilities, a user can, for example, write a **signature
+   library** simply by defining a library which has only signatures in
+   it, which allow signatures to be shared between multiple indefinite
+   libraries.
+
+2. It can be **instantiated** (possibly multiple times)
+   into a **definite library** which can be compiled. (*Extensibility*)
+   A client instantiates a library by bringing it into scope with
+   another library which exposes modules with the same names as the
+   signatures (a process called **mix-in linking**)::
+
+        library str-bytestring
+            exposed-modules: Str
+
+        library concat-bytestring
+            build-depends: str-bytestring, concat-indef
+            reexported-modules:
+                -- Concat from concat-indef is instantiated
+                -- with Str from str-bytestring.  We can
+                -- reexport it under a qualified name for
+                -- more convenient use.
+                Concat as Concat.ByteString
+
+   Compilation does not occur until the implementation of all
+   signatures is known, so that when compilation occurs, a compiler can optimize
+   as if Backpack was not present (*no performance overhead.*)
+   The types for each instantiation are considered distinct
+   (in a sense that is made precise in the `Identifiers`_ section.)
```

Comment:
You can instantiate some but not all of the holes. If you instantiate none, you have (1), if you instantiate all, it's (2). Anywhere in-between is OK, but you can only compile when you get to (2).